sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp

```diff
@@ -158,11 +159,11 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
     protocolLayerOptions.RetainUncommittedData = options.RetainUncommittedData;
     protocolLayerOptions.Close = options.Close;
     protocolLayerOptions.ContentLength = 0;
-    if (options.ContentHash.HasValue()
-        && options.ContentHash.Value().Algorithm != HashAlgorithm::Md5)
-    {
-      std::abort();
-    }
+
+    AZURE_ASSERT(
+        options.ContentHash.HasValue()
+        && options.ContentHash.Value().Algorithm == HashAlgorithm::Md5);
+
```

Comment:
whoa, this is not equivalent.

Reply:
Why not?
We are asserting that ContentHash has a value and the value needs to be Md5.  If this is NOT true, the app must crash.
The previous condition says:  If ContentHash has a value and the value is anything but Md5, abort.

Remember that ASSERT checks for `true` condition, that is, for the expected program state, in order to continue the program execution. If the assert is false, then the app aborts.

Help me understand why it is not equivalent?

Reply:
old code:

```C++
    if (options.ContentHash.HasValue()
        && options.ContentHash.Value().Algorithm != HashAlgorithm::Md5)
    {
      std::abort();
    }
```

if `ContentHash` doesn't have value, it doesn't crash.


new code:

```C++
    AZURE_ASSERT(
        options.ContentHash.HasValue()
        && options.ContentHash.Value().Algorithm == HashAlgorithm::Md5);
```

If `ContentHash` doesn't have value, it crashes

Reply:
Basically `~(A && B)` is equivalent to `~A || ~B`.

Reply:
got it.

So, I am guessing that if the ContentHash is empty, the protocol layer won't make it part of the payload as it is not a mandatory field?

In that case, what @RickWinter mentioned someplace in this PR is true, we are validating the input in the client and making a decision about if sending or not the request which goes against the SDK guidelines. The SDK should just send the values to Azure and let the server return an error to the user.  In that case, there are many abort() here which are wrong and should be removed.

Reply:
> So, I am guessing that if the ContentHash is empty, the protocol layer won't make it part of the payload as it is not a mandatory field?

Correct

Reply:
> In that case, what @RickWinter mentioned someplace in this PR is true, we are validating the input in the client and making a decision about if sending or not the request which goes against the SDK guidelines. The SDK should just send the values to Azure and let the server return an error to the user. In that case, there are many abort() here which are wrong and should be removed.

The problem is that, some APIs support both MD5 and CRC64. We don't need to do any validation on client side. While some API only support MD5, if you send out CRC64 header, server side just ignores it. It'd be very bad if we ignore some data integrity header provided by customers without telling them.